pkg/awsclient/awsclient.go

```diff
@@ -40,29 +40,42 @@ func New(config AWSClientConfig) (*AwsClient, error) {
 	return a, nil
 }
 
-func (a *AwsClient) GetAWSClientSession(ctx context.Context, clusterName, namespace string) (clientaws.ConfigProvider, error) {
-	awsCluster, err := key.GetAWSClusterByName(ctx, a.ctrlClient, clusterName, namespace)
+func (a *AwsClient) GetAWSClientSession(awsRoleARN string, region string) (clientaws.ConfigProvider, error) {
+	s, err := sessionForRegion(region)
 	if err != nil {
-		a.log.Error(err, "failed to get AWSCluster")
-		return nil, err
+		return nil, microerror.Mask(err)
 	}
+	awsClientConfig := &aws.Config{Credentials: stscreds.NewCredentials(s, awsRoleARN)}
 
-	cluster, err := capiutil.GetClusterFromMetadata(ctx, a.ctrlClient, awsCluster.ObjectMeta)
+	o, err := session.NewSession(awsClientConfig)
 	if err != nil {
-		return nil, err
+		return nil, microerror.Mask(err)
 	}
 
-	// Create the cluster scope just to reuse logic of getting proper AWS session from cluster-api-provider-aws controller code
-	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
-		Client:         a.ctrlClient,
-		Logger:         &a.log,
-		Cluster:        cluster,
-		AWSCluster:     awsCluster,
-		ControllerName: "capa-iam",
```

Comment:
I dropped the use of the function from the CAPA repository and used the same native way we do in other operators like `irsa-operator`

Reply:
I am confused about this section. I'm convinced though that we don't need to cache the sessions.

Firstly, we call `session.NewSession` twice. Once when with the cache (caching by region), and once with the `Credentials` set in the Config (but without Region). So the only session we're caching is the intial one we use to set up the credentials. As far as I can tell the reason we need the initial session is because the credentials struct returned from `stscreds.NewCredentials` will refresh the credentials, and needs an STS client to actually call the AssumeRole endpoint. Also it looks like there is no getting away with *not* creating this initial session for the Credentials struct.

Now according to the docs [here](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html#sts-endpoints) STS has a global endpoint, but it's recommended to use regional endpoints for a few reasons. However, we're always calling the AssumeRole from the MC's region; the place where capa-iam-operator is running. This means that we only need a session in _that_ region to initialize the Credentials.

Even if we decide to put the region in the second session we create (the one with the credentials) we only use these sessions to create an IAM client and an EKS client. The former isn't regioned and we anyway don't give it a region since we don't set it in the session we return. And the later sets the region when we create the client itself.

Reply:
TBH i copied this from `irsa-operator` and not think too much of it, I can drop the session part of its looks like as stupid code

Reply:
I also copied the session caching on `aws-resolver-rules-operator`, but I don't know why we needed. Does anyone know?

On that operator I use [the region of the reconciled `AWSCluster` when creating the client](https://github.com/giantswarm/aws-resolver-rules-operator/blob/main/pkg/aws/clients.go), though.

If we can't find any reason to use the cache, I'd say let's drop it.

Reply:
Ok dropping it :p